File: Programmers/[86053] 금과 은 운반하기/Johoseong/86053.py
def solution(a, b, g, s, w, t):
    start, end = 0, 4 * (10 ** 5) * (10 ** 9)
    # start, end = 0, 100
    answer = end
    N = len(g)

    while start <= end:
        cur_time = (start + end) // 2
        gold, silver, total = 0, 0, 0

        for i in range(N):
            max_move_cnt = cur_time // (t[i] * 2)
            if cur_time % (t[i] * 2) >= t[i]: # 왕복시간으로 나눈 나머지가 편도로 걸리는 시간보다 크면 +1 (편도 이동가능 의미)
                max_move_cnt += 1

            if w[i] * max_move_cnt > g[i]: # 현재 시간&도시에서, 금 최대 운송 무게
                gold += g[i]
            else:
                gold += (w[i] * max_move_cnt)
            if w[i] * max_move_cnt > s[i]: # 은 최대 운송 무게
                silver += s[i]
            else:
                silver += (w[i] * max_move_cnt)
            if w[i] * max_move_cnt > s[i] + g[i]: # 금 + 은 최대 운송 무게
                total += (s[i] + g[i])
            else:
                total += (w[i] * max_move_cnt)

        if total >= (a + b) and gold >= a and silver >= b:
            end = cur_time - 1 # 시간 더 줄임
            answer = min(answer, cur_time)
        else:
            start = cur_time + 1 # 현재 시간으론 a, b 만족 못함 -> 시간 늘림

    return answer

File: Programmers/[86053] 금과 은 운반하기/ONE/86053.java
class Solution {
    public long solution(int a, int b, int[] g, int[] s, int[] w, int[] t) {
        int n = g.length;
        long answer = (long) ((long) 1e5 * 2 * 1e9 * 2);
        long left = 0;
        long right = answer;

        while (left <= right) {
            long mid = (left + right) / 2;

            long gold = 0;
            long silver = 0;
            long sum = 0;
            long totalWeight;

            for (int i = 0; i < n; i++) {
                long moveCount = (long) Math.ceil(((double) (mid / t[i]) / 2));
                totalWeight = moveCount * w[i];
                gold += Math.min(g[i], totalWeight);
                silver += Math.min(s[i], totalWeight);
                sum += Math.min(g[i] + s[i], totalWeight);
            }

            if (gold >= a && silver >= b && sum >= a + b) {
                right = mid - 1;
                answer = Math.min(mid, answer);
            }
            else
                left = mid + 1;
        }
        return answer;
    }
}